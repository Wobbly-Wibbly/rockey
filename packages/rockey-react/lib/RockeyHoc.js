import React from 'react';
import isString from 'lodash/isString';
import classnames from 'classnames';

import rule from 'rockey/rule';

const WAS_CALLED_AS_CSS = 'WAS_CALLED_AS_CSS';
const WAS_CALLED_AS_CSS_EXTENDS = 'WAS_CALLED_AS_CSS_EXTENDS';
const WAS_CALLED_AS_REACT_COMPONENT = 'WAS_CALLED_AS_REACT_COMPONENT';

const getCallType = (...args) => {
  if (args.length === 1 && isString(args[0])) {
    return WAS_CALLED_AS_CSS_EXTENDS;
  } else if (Array.isArray(args[0])) {
    return WAS_CALLED_AS_CSS;
  } else {
    return WAS_CALLED_AS_REACT_COMPONENT;
  }
};

const renderReactComopnent = (
  BaseComponent,
  {
    props,
    css,
    displayName,
  }
) => {
  const classList = css ? css.getClassList(props) : {};

  return React.createElement(BaseComponent, {
    ...props,
    className: classnames(classList[displayName], props.className),
  });
};

let anonymysRockeyCounter = 0;

const uniq = new Map();
const getUniqKey = (BaseComponent, strings, values) => {
  return JSON.stringify([
    BaseComponent.toString(),
    ...strings,
    ...(values || []),
  ]);
};

export const getRockeyHoc = () => {
  const CreateRockeyHoc = (
    BaseComponent,
    {
      displayName,
      parentCss = null,
    } = {}
  ) => {
    if (isString(BaseComponent)) {
      displayName = displayName || BaseComponent;
    } else if (!parentCss) {
      displayName = displayName ||
        BaseComponent.displayName ||
        `AnonymysRockey${++anonymysRockeyCounter}`;
    } else {
      displayName = displayName || `AnonymysRockey${++anonymysRockeyCounter}`;
    }

    return (...args) => {
      const CALL_TYPE = getCallType(...args);

      if (CALL_TYPE === WAS_CALLED_AS_REACT_COMPONENT) {
        return renderReactComopnent(BaseComponent, {
          props: args[0],
          displayName,
          css: parentCss,
        });
      }

      const css = rule(...args);

      if (parentCss) {
        css.addParent(parentCss);
      }

      css.wrapWith(displayName);

      const RockeyHoc = function RockeyHoc(...args) {
        const CALL_TYPE = getCallType(...args);

        switch (CALL_TYPE) {
          case WAS_CALLED_AS_CSS_EXTENDS:
            return (childStrings, ...childValues) => {
              const childCss = rule(childStrings, ...childValues);

              return CreateRockeyHoc(BaseComponent, {
                displayName: args[0],
                parentCss: css,
              })(childStrings, ...childValues);
            };

          case WAS_CALLED_AS_CSS:
            return CreateRockeyHoc(BaseComponent, {
              displayName: 'Child' + displayName,
              parentCss: css,
            })(args[0], ...args.slice(1));

          case WAS_CALLED_AS_REACT_COMPONENT:
            return renderReactComopnent(BaseComponent, {
              props: args[0],
              displayName,
              css,
            });

          default:
            throw new Error('Wrong component call');
        }
      };

      RockeyHoc.displayName = `Rockey(${displayName})`;

      return RockeyHoc;
    };
  };

  return CreateRockeyHoc;
};

const RockeyHoc = getRockeyHoc();
export default RockeyHoc;
