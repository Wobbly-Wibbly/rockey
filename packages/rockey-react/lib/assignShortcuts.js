import isString from 'lodash/isString';
import isArray from 'lodash/isArray';

import htmlTags from './htmlTags';

const ucfirst = s => s.charAt(0).toUpperCase() + s.slice(1);
let counter = {};

const assignShortcuts = RockeyHoc => {
  for (const tag of htmlTags) {
    // ---- tag hoc lazy creation
    Object.defineProperty(RockeyHoc, tag, {
      get: () =>
        (...args) => {
          if (args.length === 1 && isString(args[0])) {
            // ---- use with defined displayName
            return RockeyHoc(tag, {
              displayName: ucfirst(args[0]),
            });
          } else if (isArray(args[0])) {
            // ---- use as anonymys tag shortcut
            if (!counter[tag]) {
              counter[tag] = 0;
            }

            return RockeyHoc(tag, {
              displayName: `Shortcut${ucfirst(tag)}${++counter[tag]}`,
            })(...args);
          } else {
            throw new Error(
              `shortcut.${tag} used as React Component but without defined styles. Use jsx syntax directly for such cases - "<${tag}>...</${tag}>"`
            );
          }
        },
      set: () => {
        throw new Error('Override default RockeyHoc shorcuts is disabled');
      },
    });
  }
};

export default assignShortcuts;
