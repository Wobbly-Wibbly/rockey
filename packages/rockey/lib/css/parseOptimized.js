/**
 * --------------------------------------------------------------
 * --------------------------------------------------------------
 * NOTE this is autogenerated module. Does not change it manually
 * --------------------------------------------------------------
 * --------------------------------------------------------------
 */
// const validateRule = (rule, value) => {};
// const autoprefixRule = (rule, value) => {};
const cache = new Map();
const parseCss = styles => {
  const stylesParts = styles
    .trim()
    .split(';')
    .map(s => s ? s.split(':') : null)
    .filter(s => !!s);
  const key = stylesParts.toString();
  if (!key.trim()) {
    return {};
  }
  if (cache.has(key)) {
    return cache.get(key);
  }
  const stylesJson = stylesParts.reduce(
    (json, value) => {
      if (value) {
        const key = value[0].trim();
        if (json[key]) {
          // css fallback values
          let prev = Array.isArray(json[key]) ? json[key] : [json[key]];
          json[value[0].trim()] = prev.concat([
            value.slice(1).join(':').trim(),
          ]);
        } else {
          json[value[0].trim()] = value.slice(1).join(':').trim();
        }
        // validateRule(value[0], json[value[0]]);
        // json[value[0].trim()] = autoprefixRule(value.slice(1).join(':').trim());
      }
      return json;
    },
    {}
  );
  cache.set(key, stylesJson);
  return stylesJson;
};
const isStartsWithModificator = raw => {
  return 0 === raw.indexOf('@media') ||
    0 === raw.indexOf('@keyframes') ||
    0 === raw.indexOf(':before') ||
    0 === raw.indexOf(':after') ||
    0 === raw.indexOf(':disabled') ||
    0 === raw.indexOf(':hover') ||
    0 === raw.indexOf(':link') ||
    0 === raw.indexOf(':active') ||
    0 === raw.indexOf(':visited') ||
    0 === raw.indexOf(':nth') ||
    0 === raw.indexOf('::first-line') ||
    0 === raw.indexOf(':not');
};
const parse = (raw, parent) => {
  let openedBrackets = 0;
  let openedModificatorBrackets = 0;
  let styles = '';
  let component = null;
  let components = {};
  let possibleModificator = 0;
  let modificators = {};
  let modificator = '';
  let current = '';
  let currentBackup = '';
  // ---
  const mixins = [];
  let possibleMixin = false;
  let mixin = '';
  // ----
  for (let symbol of raw) {
    if (!!!component && !!!modificator) {
      if (symbol === '_') {
        possibleMixin = true;
      }
      if (possibleMixin) {
        mixin += symbol;
        if (symbol === ' ') {
          possibleMixin = false;
          mixins.push(mixin.trim());
          current = current.replace(mixin.trim(), '');
          mixin = '';
        }
      }
    }
    if (!!!component && !!!modificator) {
      if (!possibleModificator && (symbol === '@' || symbol === ':')) {
        possibleModificator = true;
        currentBackup = current;
        current = '';
      }
      if (possibleModificator && (symbol === '{' || symbol === ';')) {
        if (isStartsWithModificator(current.trim())) {
          modificator = current;
          current = '';
        } else {
          current = currentBackup + current;
          currentBackup = '';
        }
        possibleModificator = false;
      }
    }
    if (
      !possibleModificator &&
      !!!modificator &&
      (symbol === '{' && openedBrackets === 0)
    ) {
      openedBrackets++;
      const parts = current.trim().split(' ');
      if (parts.length === 1) {
        component = parts[0];
      } else {
        styles += parts.slice(0, -1).join('');
        component = parts[parts.length - 1];
      }
      current = '';
      // continue;
    } else if (symbol === '{') {
      if (!!component) {
        openedBrackets++;
        if (openedBrackets !== 1) {
          current += symbol;
        }
      } else if (!!modificator) {
        openedModificatorBrackets++;
        if (openedModificatorBrackets !== 1) {
          current += symbol;
        }
      }
    } else if (symbol === '}') {
      if (!!component) {
        openedBrackets--;
      } else if (!!modificator) {
        openedModificatorBrackets--;
      }
      if (!!component) {
        if (openedBrackets === 0) {
          components[component] = parse(current, {
            parentType: 'component',
            name: component,
          });
          component = null;
          current = '';
        } else {
          current += symbol;
        }
      } else if (!!modificator) {
        if (openedModificatorBrackets === 0) {
          modificator = modificator.trim();
          modificators[modificator] = parse(current, {
            parentType: 'modificator',
            name: modificator,
          });
          modificator = '';
          current = currentBackup;
          currentBackup = '';
        } else {
          current += symbol;
        }
      }
    } else {
      current += symbol;
    }
  }
  current = currentBackup + current;
  currentBackup = '';
  if (current) {
    styles += current;
  }
  return {
    mixins,
    components,
    styles: parseCss(styles),
    modificators,
  };
};
export default inline => {
  return parse(inline.replace(/\r|\n/g, ''));
};
