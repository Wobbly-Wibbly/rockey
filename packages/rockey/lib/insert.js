import { rule, staticRule } from './rule';
import hash from './utils/hash';

const insert = (strings, ...values) => {
  const css = rule(strings, ...values);

  css.wrapWith(`Insert`);
  return (props = {}) => css.getClassList(props).Insert;
};

export const insertStatic = (strings, ...values) => {
  const css = staticRule(strings, ...values);

  css.wrapWith(`InsertStatic`);
  return css.getClassList().InsertStatic;
};

export default insert;
