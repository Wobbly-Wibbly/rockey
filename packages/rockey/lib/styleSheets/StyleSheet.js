import isArray from 'lodash/isArray';

const isContainer = key => key.indexOf('@media') === 0;
const isKeyFrames = key => key.indexOf('@keyframes') === 0;

// TODO: add className + description cache
const stringifyRule = (className, description) => {
  return `
    ${className} {
      ${Object.keys(description)
    .map(key => {
      if (isArray(description[key])) {
        return description[key].map(value => `${key}:${value};`).join('');
      } else {
        return `${key}:${description[key]};`;
      }
    })
    .join('')}
    }
  `;
};

export const stringifyRules = tree => {
  let css = '';

  for (const className of Object.keys(tree)) {
    css += stringifyRule(className, tree[className]);
  }

  return css;
};

const toArray = tree => {
  let rules = [];

  for (const selector of Object.keys(tree)) {
    let css = null;

    if (isContainer(selector)) {
      css = `
        ${selector} {
          ${stringifyRules(tree[selector])}
        }
      `;
    }
    if (isKeyFrames(selector)) {
      css = `
        ${selector} {
          ${stringifyRules(tree[selector])}
        }
      `;
    } else {
      css = stringifyRule(selector, tree[selector]);
    }

    rules.push(css);
  }

  return rules;
};

const sheetForTag = tag => {
  if (tag.sheet) {
    return tag.sheet;
  }

  // this weirdness brought to you by firefox
  for (let i = 0; i < document.styleSheets.length; i++) {
    if (document.styleSheets[i].ownerNode === tag) {
      return document.styleSheets[i];
    }
  }
};

const QUEUE_THRESHOLD = 200;

export default class StyleSheet {
  rules = [];
  mixins = [];

  timeouts = {
    insert: null,
    mixins: null,
  };

  queue = {
    rules: [],
    mixins: [],
  };

  counter = {
    rueles: 0,
    mixins: 0,
  };

  constructor(sheets) {
    this.styleNode = sheets.main;
    this.styleMixinsNode = sheets.mixins;
  }

  getRules() {
    return this.rules;
  }

  getMixins() {
    return this.mixins;
  }

  clear() {
    this.mixins = [];
    this.rules = [];

    if (this.timeouts.mixins) {
      clearTimeout(this.timeouts.mixins);
    }

    if (this.timeouts.rules) {
      clearTimeout(this.timeouts.rules);
    }

    this.styleNode.textContent = '';
    this.styleMixinsNode.textContent = '';
  }

  insertMixins(tree) {
    const rules = toArray(tree);

    this.queue.mixins = this.queue.mixins.concat(rules);

    if (this.queue.mixins.length >= 150) {
      this.styleMixinsNode.appendChild(
        document.createTextNode(this.queue.mixins.join(''))
      );

      this.mixins = this.mixins.concat(this.queue.mixins);
      this.queue.mixins.length = 0;
      return;
    }

    if (this.timeouts.mixins) {
      clearTimeout(this.timeouts.mixins);
    }

    this.timeouts.mixins = setTimeout(() => {
      this.styleMixinsNode.appendChild(
        document.createTextNode(this.queue.mixins.join(''))
      );

      this.mixins = this.mixins.concat(this.queue.mixins);
      this.queue.mixins.length = 0;
    });
  }

  insertRules(tree) {
    const rules = toArray(tree);
    this.queue.rules = this.queue.rules.concat(rules);

    if (this.queue.rules.length >= QUEUE_THRESHOLD) {
      this.styleNode.appendChild(
        document.createTextNode(this.queue.rules.join(''))
      );

      this.rules = this.rules.concat(this.queue.rules);
      this.queue.rules.length = 0;
      return;
    }

    if (this.timeouts.insert) {
      clearTimeout(this.timeouts.insert);
    }

    this.timeouts.insert = setTimeout(() => {
      this.styleNode.appendChild(
        document.createTextNode(this.queue.rules.join(''))
      );

      this.rules = this.rules.concat(this.queue.rules);
      this.queue.rules.length = 0;
    });
  }
}
