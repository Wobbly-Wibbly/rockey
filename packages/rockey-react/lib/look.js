import rule from 'rockey/rule';
import when from 'rockey/when';
import insert from 'rockey/insert';
import condition from 'rockey/condition';

import htmlTags from './htmlTags';

import RockeyHoc from './';

const look = BaseComponent =>
  (...args) => {
    const css = rule(...args);

    return css.transform((tree, create) => {
      const components = Object.keys(tree.components);
      const size = components.length;

      if (!size) {
        throw new Error(
          'Rockey look should containt at least one root component'
        );
      }

      const parentDisplayName = components[0];
      const parentCss = create({
        components: {
          [parentDisplayName]: tree.components[parentDisplayName],
        },
      });

      const children = {};

      for (let i = 1; i < size; i++) {
        const displayName = components[i];

        const comopnentCss = create({
          components: {
            [displayName]: tree.components[displayName],
          },
        });
        comopnentCss.addParent(parentCss);

        children[displayName] = RockeyHoc(BaseComponent, {
          displayName,
          parentCss: comopnentCss,
        });
      }

      return {
        ...children,
        [parentDisplayName]: RockeyHoc(BaseComponent, {
          displayName: parentDisplayName,
          parentCss: parentCss,
        }),
      };
    });
  };

look.when = when;
look.condition = condition;
look.insert = insert;

for (const tag of htmlTags) {
  // ---- tag hoc lazy creation
  Object.defineProperty(look, tag, {
    get: () => (...args) => look(tag)(...args),
  });
}

export default look;
