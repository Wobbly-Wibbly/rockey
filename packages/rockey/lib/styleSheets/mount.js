import StyleSheet from './StyleSheet';

const isBrowser = typeof document !== 'undefined';

const createStyleSheet = type => {
  if (!isBrowser) {
    return {
      type,
      innerHtml: '',
    };
  } else {
    const mountNode = document.createElement('style');

    mountNode.type = 'text/css';

    document.head.appendChild(mountNode);
    mountNode.textContent = '';

    return mountNode;
  }
};

const mount = () => {
  return new StyleSheet({
    main: createStyleSheet('main'),
    mixins: createStyleSheet('mixins'),
  });
};

export default mount;
