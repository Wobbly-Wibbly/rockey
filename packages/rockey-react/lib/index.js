import when from 'rockey/when';
import insert from 'rockey/insert';
import condition from 'rockey/condition';

import { getRockeyHoc } from './RockeyHoc';
import assignShorcuts from './assignShorcuts';
import look from './look';

const RootRockeyHoc = getRockeyHoc();

RootRockeyHoc.look = look;
RootRockeyHoc.when = when;
RootRockeyHoc.condition = condition;
RootRockeyHoc.insert = insert;

assignShorcuts(RootRockeyHoc);

export default RootRockeyHoc;
