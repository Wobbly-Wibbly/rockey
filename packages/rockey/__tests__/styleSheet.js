import rule, { staticRule, clearStylesCache } from '../lib/rule';
import when from '../lib/when';
import { getClassName } from '../lib/css/getClassName';
import { getRules, getMixins, clearStyles } from '../lib/styleSheets';

jest.mock('../lib/utils/hash', () => {
  return () => '{{ hash }}';
});

console.warn = jest.fn();

const primary = when(props => props.isPrimary)`
  color: blue
`;

test('inserted styles', () => {
  clearStyles();
  clearStylesCache();

  const css = rule`
    Button {
      color: red;

      ${primary}
    }
  `;

  expect(css.getClassList()).toEqual({
    Button: ['Button-{{ hash }}'],
  });
  expect(getRules()).toEqual([]);
  expect(getMixins()).toEqual([]);

  return new Promise((resolve, reject) => {
    setTimeout(() => {
      try {
        expect(getRules()).toMatchSnapshot();

        expect(getMixins()).toMatchSnapshot();

        resolve();
      } catch (e) {
        reject(e);
      }
    });
  });
});

test('inserted styles with mixin', () => {
  clearStyles();
  clearStylesCache();

  const css = rule`
    Button {
      color: purple;

      ${primary}
    }
  `;

  expect(
    css.getClassList({
      isPrimary: true,
    })
  ).toEqual({
    Button: ['Button-{{ hash }}', 'Mixin-anonWhen-{{ hash }}-1'],
  });

  expect(getRules()).toEqual([]);
  expect(getMixins()).toEqual([]);

  return new Promise((resolve, reject) => {
    setTimeout(() => {
      try {
        expect(getRules()).toMatchSnapshot();

        expect(getMixins()).toMatchSnapshot();

        resolve();
      } catch (e) {
        reject(e);
      }
    });
  });
});

test('inserted above threshold', () => {
  clearStyles();
  clearStylesCache();

  const styles = [];

  for (let i = 0; i <= 200; i++) {
    styles.push(
      `
      Button${i} {
        color: purple;
      }
    `
    );
  }

  const css = rule`${styles.join('')}`;

  css.getClassList();

  expect(getRules().length).toBeGreaterThan(0);
  expect(getMixins().length).toEqual(0);

  expect(getRules()).toMatchSnapshot();

  expect(getMixins()).toMatchSnapshot();
});
