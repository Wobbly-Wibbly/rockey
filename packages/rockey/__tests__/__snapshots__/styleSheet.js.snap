// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inserted above threshold 1`] = `
Array [
  "
    .Button0-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button1-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button2-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button3-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button4-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button5-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button6-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button7-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button8-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button9-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button10-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button11-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button12-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button13-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button14-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button15-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button16-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button17-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button18-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button19-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button20-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button21-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button22-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button23-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button24-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button25-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button26-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button27-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button28-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button29-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button30-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button31-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button32-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button33-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button34-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button35-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button36-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button37-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button38-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button39-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button40-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button41-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button42-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button43-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button44-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button45-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button46-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button47-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button48-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button49-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button50-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button51-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button52-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button53-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button54-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button55-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button56-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button57-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button58-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button59-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button60-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button61-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button62-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button63-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button64-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button65-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button66-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button67-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button68-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button69-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button70-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button71-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button72-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button73-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button74-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button75-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button76-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button77-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button78-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button79-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button80-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button81-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button82-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button83-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button84-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button85-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button86-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button87-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button88-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button89-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button90-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button91-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button92-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button93-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button94-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button95-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button96-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button97-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button98-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button99-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button100-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button101-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button102-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button103-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button104-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button105-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button106-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button107-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button108-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button109-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button110-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button111-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button112-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button113-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button114-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button115-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button116-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button117-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button118-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button119-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button120-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button121-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button122-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button123-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button124-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button125-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button126-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button127-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button128-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button129-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button130-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button131-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button132-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button133-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button134-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button135-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button136-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button137-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button138-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button139-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button140-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button141-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button142-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button143-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button144-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button145-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button146-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button147-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button148-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button149-{{ hash }} {
      color:purple;
    }
  ",
  "
    .Button150-{{ hash }} {
      color:purple;
    }
  ",
]
`;

exports[`inserted above threshold 2`] = `Array []`;

exports[`inserted styles 1`] = `
Array [
  "
    .Button-{{ hash }} {
      color:red;
    }
  ",
]
`;

exports[`inserted styles 2`] = `Array []`;

exports[`inserted styles with mixin 1`] = `
Array [
  "
    .Button-{{ hash }} {
      color:purple;
    }
  ",
]
`;

exports[`inserted styles with mixin 2`] = `
Array [
  "
    .Mixin-anonWhen-{{ hash }}-1.Button-{{ hash }} {
      color:blue;
    }
  ",
]
`;
